runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -195,6 +195,8 @@
   // retrieving extra work from Windmill without working on it, leading to better
   // prioritization / utilization.
   static final int MAX_WORK_UNITS_QUEUED = 100;
+  // Maximum bytes of WorkItems being processed in the work queue at a time.
+  static final int MAX_WORK_UNITS_BYTES = 500 << 20; // 500MB
```

Comment:
This would only throttle pipelines that are successfully operating with >500 MB outstanding at a time and recovering from this scenario faster than windmill could deliver new data to the worker. I'd expect this to be a very, very small percentage of pipelines, effectively only those whose work items were consistently about (NON_OOM_USABLE_MEMORY / 400) in size, which would be rare for them to be so precisely tuned to that, but also rare that they would be able to process 500 MB of data before windmill could deliver them more.

Reply:
Still, making it a pipelineoption will allow much more agile adjustments by the service.

Reply:
Following that logic `MAX_WORK_UNITS_QUEUED` should be configurable as well then. Just like how having a memory limit can throttle performance if the work units are big, having a count limit when the work units are small does the same.

Reply:
Done.

Reply:
The argument that pipelines would always OOM if they exceeded 500MB active breaks down due to pipelines with limited parallelism. Streaming dataflow does not return additional work for keys that are already active on the worker.

For example, if you have a pipeline writing to files with fixed shards and there are only 8 shards per worker, the current queued/active is 800MB (due to 100MB work bundle limit).  Since there is by default ~4GB and configurably a lot more memory for the java harness such pipelines would not be ooming.

So I'm still concerned that setting the default to 500MB could limit pipelines that previously worked.  Now that it is an option it is at least something we can tune but it still seems like it could cause issues and should perhaps be scaled based upon machine memory.

Reply:
For elements we have separate limits for the queue and for active (since there are a limited # of active threads each processing one work item).

Should we have the distinction between queued and active bytes? Queued bytes seems like it would be safer to have a default lower size, where active bytes might throttle pipelines that currently work.

Reply:
What would your preferred default limit be? The default memory limit for dataflow streaming engine workers is 8G (n1-standard-2), would you prefer a 2G or 4G default for such machines?

If you'd prefer this to be scaled on available jvm memory, what would the preferred default fraction be?

> For elements we have separate limits for the queue and for active (since there are a limited # of active threads each processing one work item).

We don't really. We have a number of threads and a number of queue slots, so threads + queue slots = outstanding. With the default of 100 work threads, if we didnt' limit active memory, we would still ahve this problem.

Reply:
This is now defaulted to 50% of JVM max mem